### ISSUE TITLE
Convert parsed AST to AST Builder

### ISSUE THREAD

**Issue by cemoktra** [gluesql/gluesql#1405]
Hi,

i have a use case that would require extending a parsed query. If i read the docs correctly when parsing SQL i get an AST of the `ast` module. To use the builder i need the `ast_builder` module. Is there a way to convert them already that overlooked? Is this possible?

**Comment by ever0de** [comment]
Do you mean the conversion from `AST` to `AstBuilder`, which is already parsed?

**Comment by cemoktra** [comment]
Yes

**Comment by ever0de** [comment]
How do you intend to use it?

I'm not sure what's going on, but the output of `AstBuilder` is `AST`.
What do you mean by converting from `AST` to `AstBuilder`?
For example, the
- Do you expect creating an AST to generate AstBuilder rust code?

**Comment by cemoktra** [comment]
I parse SQL, use AST builder to extend it, then output SQL again

**Comment by ever0de** [comment]
In an AST extended with AstBuilder, you can use `ToSql::to_sql` to get a sql string.
Is this what we want? 👀

**Comment by cemoktra** [comment]
Yes, but i was asking for the conversion between AST and AstBuilder. E.g.:

- i have a query `SELECT a FROM t`
- this query is checked at compiled by `sqlx`
- now i parse that query into an AST
- convert this AST into an AST builder
- add dynamic where clauses using the AST builder
- execute final sql

**Comment by ever0de** [comment]
Oh okay. I get it.

What you said is partially possible. 
- By "partially" I mean the `AstNodes that are implemented.
- See `core/src/ast_builder::*` for all implementations.

For example, an `ExprNode` can be extended with the `Expr` variant because it exists.

**Comment by panarch** [comment]
two points

> convert this AST into an AST builder

Currently you can directly execute the AST rather than converting that into AST builder.

> add dynamic where clauses using the AST builder

though we don't have AST mutable visitor yet, but you can still change the AST by using mutable borrow.

---

AST Builder is for making AST so.. there's no certain path from AST to AST Builder now.
But AST Builder accepts the AST nodes as a parameter.

e.g.
https://github.com/gluesql/gluesql/blob/deb56a429a275520f6c4e7354609f86c247550e1/core/src/ast_builder/expr/mod.rs#L40-L44

In `ExprNode`, you can make `ExprNode` by using already built `ast::Expr` because `ExprNode` has `ExprNode::Expr(Cow<'a, Expr>)` as a variant.
